core/src/main/java/com/orientechnologies/orient/core/storage/impl/local/OAbstractPaginatedStorage.java

```diff
@@ -1466,7 +1538,9 @@ public int getClusterIdByName(final String clusterName) {
             for (ORecordOperation txEntry : newRecords) {
               ORecord rec = txEntry.getRecord();
 
-              if (rec.isDirty()) {
+              if (allocated && rec.getIdentity().isPersistent()) {
+                positions.put(txEntry, new OPhysicalPosition(rec.getIdentity().getClusterPosition()));
+              } else if (rec.isDirty() && !rec.getIdentity().isPersistent()) {
```

Comment:
If the `rid` is persisted, we get some strange behavior here. The record will not receive its entry in `positions`, its physical position will be `null` in `commitEntry` call, that will trigger record update logic instead of create. Not sure is it intended behavior or not. If not, it's better to add an assertion here to postulate the expected state and behavior.

Reply:
This logic is the complex part, it has three behaviours, when the record is allocated ahead [1], when is not allocated[2], when the id is generated by another node[3]. 

1) when the record is allocated ahead we should put the entry in positions and that will trigger an update as you noticed, this update will change the record from allocated to persistent.

2) when the record is not allocated and i've the rid not persistent, this happen in the normal tx flow.

3) The rid is persistent but not allocated this will force a creation of a rid. this happen when a node apply a transaction from another node, the record created by the other node don't exist but have valid rid.

So from my concern here all the case are valid, is not too clear but is the best that we can get on the current storage architecture.

Reply:
@tglman  if you think that all cases are valid, could you add the same message which you put there as comment in code, so other developers will not scratch their head thinking what is going on ))